### ISSUE TITLE
[QUESTION]: Inconsistent results across different machines/graphics cards

### ISSUE THREAD

**Issue by YoochanMyung** [chemprop/chemprop#534]
**What are you trying to do?**
: The performance is not consistent if I run same inputs/settings on different machines/graphics cards.

So, I have been running hyperparameter optimization across different machine, for example, A, B and C.
1. If I use the best arguments and hyperparameters obtained from A machine and run a prediction to get a model on other machines, the performance changes quite a lot.
2. Even if I stick with the same machine, A, but choose different graphics card (I have H100 and A100), the performance varies based on the graphic card.  (As the following results)

**Previous attempts**
I have been testing on the chemprop  1.6.1, numpy 1.24.3, torch 2.1.0, and I will provide further if it's required.

**Screenshots**
```
=====[Using H100]======
3-fold cross validation
    Seed 0 ==> test mcc = 0.823887
    Seed 0 ==> test f1 = 0.936508
    Seed 0 ==> test auc = 0.950158
    Seed 0 ==> test accuracy = 0.918782
    Seed 1 ==> test mcc = 0.845373
    Seed 1 ==> test f1 = 0.944664
    Seed 1 ==> test auc = 0.954475
    Seed 1 ==> test accuracy = 0.928934
    Seed 2 ==> test mcc = 0.823284
    Seed 2 ==> test f1 = 0.936759
    Seed 2 ==> test auc = 0.952261
    Seed 2 ==> test accuracy = 0.918782
Overall test mcc = 0.830848 +/- 0.010274
Overall test f1 = 0.939310 +/- 0.003787
Overall test auc = 0.952298 +/- 0.001763
Overall test accuracy = 0.922166 +/- 0.004786
Elapsed time = 0:04:29
====================

=====[Using A100]======
3-fold cross validation
    Seed 0 ==> test mcc = 0.829087
    Seed 0 ==> test f1 = 0.938614
    Seed 0 ==> test auc = 0.953004
    Seed 0 ==> test accuracy = 0.921320
    Seed 1 ==> test mcc = 0.840716
    Seed 1 ==> test f1 = 0.942346
    Seed 1 ==> test auc = 0.955176
    Seed 1 ==> test accuracy = 0.926396
    Seed 2 ==> test mcc = 0.836325
    Seed 2 ==> test f1 = 0.940000
    Seed 2 ==> test auc = 0.950761
    Seed 2 ==> test accuracy = 0.923858
Overall test mcc = 0.835376 +/- 0.004795
Overall test f1 = 0.940320 +/- 0.001540
Overall test auc = 0.952980 +/- 0.001803
Overall test accuracy = 0.923858 +/- 0.002072
Elapsed time = 0:04:50
====================
```
Best,

**Comment by JacksonBurns** [comment]
Hi @YoochanMyung thanks for the thorough report!

This is just a hardware limitation related to PyTorch. Different GPU architectures are slightly different, which leads to different mathematical results when doing floating point operations. You can see discussion on the PyTorch forum [here](https://discuss.pytorch.org/t/different-machines-different-results/100126). We already set the random seed before training by default to take care of software reproducibility, so this is beyond our reach.

The good news, though, is that these values are not significantly different (a t-test gives a p-value of ~0.6 for the accuracy metric, for example). `chemprop` is working as intended.
